examples/stream/server.c

```diff
@@ -0,0 +1,232 @@
+/*= -*- c-basic-offset: 4; indent-tabs-mode: nil; -*-
+ *
+ * librsync -- the library for network deltas
+ *
+ * Copyright (C) 2024 by Lars Erik Wik <lars.erik.wik@northern.tech>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+#include <librsync.h>
+
+#include "common.h"
+
+
+static char in_buf[BUFFER_SIZE], out_buf[BUFFER_SIZE];
+
+static int accept_connection(void);
+static int recv_signature(int sock, rs_signature_t **sig);
+static int send_delta(int sock, rs_signature_t *sig, const char *fname);
+
+int main(int argc, char *argv[])
+{
+    /* Parse arguments */
+    if (argc < 2) {
+        printf("USAGE: %s <FILENAME>", argv[0]);
+        return EXIT_FAILURE;
+    }
+    const char *fname = argv[1];
+
+    puts("Waiting for connection...");
+    int sock = accept_connection();
+    if (sock == -1) {
+        return EXIT_FAILURE;
+    }
+
+    puts("Receiving signature...");
+    rs_signature_t *sig;
+    int ret = recv_signature(sock, &sig);
+    if (ret == -1) {
+        close(sock);
+        return EXIT_FAILURE;
+    }
+
+    puts("Sending delta...");
+    ret = send_delta(sock, sig, fname);
+    rs_free_sumset(sig);
+    if (ret == -1) {
+        close(sock);
+        return EXIT_FAILURE;
+    }
+
+    puts("Success!");
+    return EXIT_SUCCESS;
+}
+
+static int accept_connection(void) {
+    /* Create socket */
+    int sock = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock == -1) {
+        perror("Failed to create socket");
+        return -1;
+    }
+
+    /* Enable reuse address */
+    int opt = 1;
+    setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));
+
+    /* Assign IP address and port */
+    struct sockaddr_in server_addr = { 0 };
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = htons(INADDR_ANY);
+    server_addr.sin_port = htons(PORT);
+
+    /* Bind socket to given IP address */
+    int ret = bind(sock, (struct sockaddr *)&server_addr, sizeof(server_addr));
+    if (ret == -1) {
+        perror("Failed to bind socket");
+        close(sock);
+        return -1;
+    }
+
+    /* Listen for incoming connections */
+    ret = listen(sock, 1);
```

Comment:
`1` is a very short queue/backlog. Not that it matters here, but you never know who reuses the code for something else ;)

Reply:
Yeah, I also `close()` the listening socket immediately after accepting the connection. I just wanted to keep the non-`librsync` code as minimal as possible. But it at least deserves a comment with a warning.